### ISSUE TITLE
remove Rec("a" -> 1) syntax in favor of Rec(a = 1)

### ISSUE THREAD

**Issue by cvogt** [scala-records/scala-records#118]
@vjovanov @gzm0 what do you think about removing this syntax? Both are supported right now. The second one highlights the type-safety and looks like case class constructors. And no need to have 2 apis.

P.S.: Slowly finding time again for records. Investigating the differences between cvogt/compossible and scala-records as preparation for a future merge.

**Comment by gzm0** [comment]
I think this syntax can be useful if there are "weird" names involved. I prefer to write:

```scala
Rec("=" -> 1)
```

than

```scala
Rec(`=` = 1)
```

Also, `applyDynamic` does not always work fully. This allows for a fallback when it fails (we maybe even should have the equivalent `apply` syntax btw.).

**Comment by cvogt** [comment]
Interesting.

If normally `Rec(foo = 1)` is encouraged, then
    Rec(`=` = 1)
still highlights the type-safety and for a case class you would also write
    Constr(`=` = 1)
I think small things like this and not diverging from following how existing syntax does it, go a long way in communicating things about behavior at first sight.

When does applyDynamicNamed fail?

**Comment by gzm0** [comment]
> when does applyDynamicNamed fail

I don't have a particular case at hand, but `Dynamic` in general does not have the reputation to be the stablest part of scalac... One of the concerns is also how it interacts with name mangling...

But it is true that the syntax is probably redundant (and "weird" names should not be used in the first place)...

@vjovanov why did we introduce this in the first place?

**Comment by gzm0** [comment]
Related to #110

**Comment by gzm0** [comment]
Hmm... that might **be** your use case...

**Comment by cvogt** [comment]
#110 looks like a fixable bug. The method names/map keys need to be en/decoded scalac names. Doesn't seem like we need the `"key" -> value" syntax for that.

**Comment by gzm0** [comment]
You are probably right. @vjovanov what do you think of getting rid of the string syntax either?

@cvogt could you please go through our tests and see if there is no case that we really can't do with backticks? (for example, empty fields)

We **must** assume that field names are arbitrary (if coming from another backend). So even if we find a single case where backtick doesn't work, we must retain the syntax.

One concern there might be, is that `applyDynamic` already gives us the encoded name. Since encoding is not reversible, that would be a complete showstopper...

@cvogt if you can test these couple of points (and supply tests for them in the PR), I'm ok with removing the syntax (given @vjovanov is OK, too).

**Comment by cvogt** [comment]
@gzm0 ok, I will :)

**Comment by vjovanov** [comment]
Sorry for the delay, my mailbox does not get empty these days; I wanted to do that long time ago. The arrow syntax also does not work with pattern matching. In this case we should also remove the tuple syntax and also remove both syntaxes in pattern matching. So record creation: 
```
  Rec(foo = 1)
```
and extraction: 
```
  case Rec(foo) => 
```

Quite elegant I would say.

**Comment by gzm0** [comment]
Ah.... I remember now: If you want to have patterns on the contents, you need the tuple syntax:

```scala
case Rec(("foo", x: String)) =>
```

**Comment by gzm0** [comment]
Actually, that one works. But this one does not:

```scala
case Rec(("x", Some(x))) =>
```

**Comment by vjovanov** [comment]
Yes, the arrow also does not work in pattern matching. For values we had `x @ 1` if you need to match the exact value. I would just stay with that and drop all other complexity. 

It would be good to keep the library small and to have one way to do things.

**Comment by cvogt** [comment]
what about @gzm0's example? Could we do `x @ Some(y)`, where x is the record field and y is bound from the value?

I am still not very happy with the fact that x is a constraint on the name, not a binding the way pattern matching is implemented for records. But I don't have a better suggestion either at the moment.

**Comment by vjovanov** [comment]
This one works as well. I am not happy as well :) But I do not see what can we do...

**Comment by gzm0** [comment]
Ah, yes. Indeed. Name binding works in this case. But what still doesn't, is if you have a field with name `1`. You can't use backtick syntax in this case, since it has special meaning in pattern matching.

**Comment by gzm0** [comment]
Btw. we can make the arrow work if we allow and additional import (bascially a `object ->` with an `unapply`).

**Comment by vjovanov** [comment]
This is also a blocker for the `@` approach; it is anyhow totally non-intuitive. 

So we have serious blockers for all non-string-field-names solutions. How about we use one syntax and stick to it. E.g.,
```
 Rec(`field -> 1)
```
or even
```
 Rec(`field := 1)
```
as `->` is extremely annoying to type on QWERTY. 

This would work for all and it is not so bad for typing. Records are anyhow mostly created in queries.

**Comment by gzm0** [comment]
Single back-tick? Was that a glitch? I think we should stick to existing tuple and named apply syntaxes and not introduce anything additional if we don't absolutely have to.

I think the discussion we are having shows that we need strings as a backup, but named apply should be preferred. Also, I actually like the approach we currently have in pattern matching. While it is not strictly aligned with how pattern matching works in Scala otherwise, it is IMO something that makes sense.

**Comment by gzm0** [comment]
Honestly, I think it is best to close this and stick with the syntax we have. I really don't think any alternative we have found is so much better that it would justify a change...

**Comment by vjovanov** [comment]
@gzm0 Updated the comment, I wanted to use single apostrophe.

How about we bless the syntax we want as the main one, and leave the strings in appendix for those that hit the issue. If somebody uses tuples/arrows without a good reason (i.e., no decoding needed) we issue a warning in compilation.

**Comment by vjovanov** [comment]
I would only decode field names by default as `foo-bar` occurs often in databases and `foo$minusbar` never.

**Comment by gzm0** [comment]
@vjovanov Yes, I agree decoding field names in named apply is the right way. Also since it is close to the Scala source code.

@vjovanov The symbol syntax, doesn't it also have issues with special characters?

>  If somebody uses tuples/arrows without a good reason (i.e., no decoding needed) we issue a warning in compilation.

Not a fan (except if there is a way to turn it off). We shouldn't force users to write this:

```scala
val x = Rec(foo = 1, bar = 1) // IMO choosing arrow syntax for consistency should be allowed here
val y = Rec("foo$" -> 1, "bar" -> 2)
```

**Comment by gzm0** [comment]
So I really think we shouldn't change anything in the code, but just document that named apply is the preferred way to construct records...

**Comment by vjovanov** [comment]
When you put it like this I am not a fan either :)

@cvogt is this OK with you?

**Comment by vjovanov** [comment]
Can we have the `=>` object imported by default for pattern matching? Also, how about we drop the arrow completely? If we need this feature only for backup why would we provide fancy syntax for the backup?

**Comment by gzm0** [comment]
The arrow in creation is just different Scala syntax for the arrow. We don't do anything about it. We currently don't have the `->` object for pattern matching. We could just create it (but the import shouldn't be automatic IMO).

**Comment by vjovanov** [comment]
We do treat the arrow [specially](https://github.com/scala-records/scala-records/blob/master/core/src/main/scala/records/CommonMacros.scala#L67).
IMO we should just not support it as it makes it easier to use something we really do not want people to use.

**Comment by gzm0** [comment]
Ahhhh... I hoped I could sneak this little one past. It is discussable whether we should support it or not indeed. But I think to the average programmer, `(a, b)` and `a -> b` are equally magic syntaxes for tuples (which both are not strictly speaking literals).

By your argument, we should also disallow the `apply` method of `Tuple2` and have people write this:

```scala
Rec(new Tuple2("name", "Hans"))
```

**Comment by vjovanov** [comment]
OK, I buy it. For the string literals we should compatible with the Scala Standard library both for construction and deconstruction. So construction would be
```
Rec(("foo", 1))
Rec(new Tuple2("foo", 1))
Rec(new Tuple2[String, Int]("foo", 1)) 
Rec(Tuple2("foo", 1))
Rec(Tuple2[String, Int]("foo", 1))
Rec("foo" -> 1)
Rec("foo" → 1)
``` 
and deconstruction
```
case Rec(("foo", 1)) => ...
case Rec(Tuple2("foo", 1)) => ...
```

And then we change the documentation. Any further ideas?

**Comment by gzm0** [comment]
I think that's good.

**Comment by cvogt** [comment]
fine with me. One suggestion however, how about we change the tests to mainly use the applyDynamicNamed syntax, so that people who use them as a reference use that syntax by default?

**Comment by vjovanov** [comment]
Good idea, both documentation and tests.

**Comment by cvogt** [comment]
awesome, thank you for the very healthy discussion!

**Comment by cvogt** [comment]
updated title and description to cover conclusing

**Comment by gzm0** [comment]
Even after disallowing `$`, (see #110), we sometimes need the tuple syntax:

- Pattern matching
- Other weird names:

    ```scala
    Rec(`-` = 1)
    // <console>:1: error: illegal start of simple expression
    ```

**Comment by cvogt** [comment]
Pattern matching, because `-backticks have different semantics there and while normally they can be avoided, for Records it may be too much trouble.

**Comment by cvogt** [comment]
In this use case using a record as a syntax for a mongodb query:
```scala
MongoCollection("...").find( Json.toJson( Record( name = Record( `$eq` = "chris" ) ) ) )
```
using applyDynamicNamed is actually not type-safe. Maybe we should advertise the tuple syntax as preferred for these cases.
```scala
MongoCollection("...").find( Json.toJson( Record( "name" -> Record( "$eq" -> "chris" ) ) ) )
```

**Comment by gzm0** [comment]
`applyDynamicNamed` is only not typesafe because we do not always agree internally whether a string needs to be decoded or not. But we should definitely test that we can have records that contain a key "-" and a key "$minus" at the same time.

**Comment by cvogt** [comment]
that too, but in the above use case the fields are never used in scala, records are only used as builders for a serialization format. that's why it is not type-safe.

**Comment by gzm0** [comment]
Oh... I don't think that has anything to do with records... `MongoCollection("...").find` is inherently un-typesafe.

**Comment by cvogt** [comment]
yes, but in this use case you create a type-safe structure only to use it dynamically and ignore all type-safety. Encouraging the tuple construction syntax could help communicate that by convention.

**Comment by gzm0** [comment]
I think we should leave that up to each individual project... But yes, in general that might make sense

**Comment by cvogt** [comment]
I think an authoritative recommendation can be helpful :). But not enforcing it.
